### ISSUE TITLE
[SALT REPO] rhel 7 and rhel5 are failing to install

### ISSUE THREAD

**Issue by gtmanfred** [saltstack/salt#29094]
```
[root@rhel7-2-dfw-cloudqe4 ~]# yum install -y python-jinja2
Loaded plugins: langpacks, product-id, rhnplugin, search-disabled-repos, subscription-manager
This system is receiving updates from RHN Classic or Red Hat Satellite.
Resolving Dependencies
--> Running transaction check
---> Package python-jinja2.noarch 0:2.7.2-2.el7 will be installed
--> Processing Dependency: python-babel >= 0.8 for package: python-jinja2-2.7.2-2.el7.noarch
--> Running transaction check
---> Package python-babel.noarch 0:0.9.6-8.el7 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

============================================================================================================================================================================================================
 Package                                           Arch                                       Version                                          Repository                                              Size
============================================================================================================================================================================================================
Installing:
 python-jinja2                                     noarch                                     2.7.2-2.el7                                      repo-saltstack-el7                                     515 k
Installing for dependencies:
 python-babel                                      noarch                                     0.9.6-8.el7                                      repo-saltstack-el7                                     1.4 M

Transaction Summary
============================================================================================================================================================================================================
Install  1 Package (+1 Dependent package)

Total size: 1.9 M
Installed size: 8.1 M
Downloading packages:
warning: /var/cache/yum/x86_64/7Server/repo-saltstack-el7/packages/python-babel-0.9.6-8.el7.noarch.rpm: Header V3 RSA/SHA256 Signature, key ID f4a80eb5: NOKEY
Retrieving key from https://repo.saltstack.com/yum/rhel7/SALTSTACK-GPG-KEY.pub


The GPG keys listed for the "SaltStack EL7 Repo" repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.


 Failing package is: python-babel-0.9.6-8.el7.noarch
 GPG Keys are configured as: https://repo.saltstack.com/yum/rhel7/SALTSTACK-GPG-KEY.pub
```

This is the error on RHEL 7, looks like you aren't correctly signing the packages in the repository.

and on RHEL5, it is depending on i686 versions of things? so when the package is tried to install, the python26.x86_64 gets pulled in, but not python26-libs.x86_64, causing the deploy to fail.

```
[root@rhel5-2-dfw-cloudqe4 ~]# yum install python26
Loaded plugins: rhnplugin
This system is receiving updates from RHN Classic or RHN Satellite.
Setting up Install Process
Resolving Dependencies
--> Running transaction check
---> Package python26.i386 0:2.6.8-3.el5 set to be updated
--> Processing Dependency: libpython2.6.so.1.0 for package: python26
--> Processing Dependency: python26-libs for package: python26
---> Package python26.x86_64 0:2.6.8-3.el5 set to be updated
--> Running transaction check
---> Package python26-libs.i386 0:2.6.8-3.el5 set to be updated
--> Finished Dependency Resolution

Dependencies Resolved

============================================================================================================================================================================================================
 Package                                           Arch                                       Version                                          Repository                                              Size
============================================================================================================================================================================================================
Installing:
 python26                                          i386                                       2.6.8-3.el5                                      repo-saltstack-el5                                     6.5 M
 python26                                          x86_64                                     2.6.8-3.el5                                      repo-saltstack-el5                                     6.5 M
Installing for dependencies:
 python26-libs                                     i386                                       2.6.8-3.el5                                      repo-saltstack-el5                                     671 k

Transaction Summary
============================================================================================================================================================================================================
Install       3 Package(s)
Upgrade       0 Package(s)

Total download size: 14 M
Is this ok [y/N]: y
Downloading Packages:
(1/3): python26-libs-2.6.8-3.el5.i386.rpm                                                                                                                                            | 671 kB     00:00
(2/3): python26-2.6.8-3.el5.i386.rpm                                                                                                                                                 | 6.5 MB     00:01
(3/3): python26-2.6.8-3.el5.x86_64.rpm                                                                                                                                               | 6.5 MB     00:01
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                       3.3 MB/s |  14 MB     00:04
Running rpm_check_debug
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
error: rpmdbNextIterator: skipping h#     198 Header V4 RSA/SHA1 signature: BAD, key ID e782684e
ERROR with rpm_check_debug vs depsolve:
libpython2.6.so.1.0()(64bit) is needed by python26-2.6.8-3.el5.x86_64
Complete!
(1, [u'Please report this error in https://bugzilla.redhat.com/enter_bug.cgi?product=Red%20Hat%20Enterprise%20Linux%205&component=yum'])
```

**Comment by gtmanfred** [comment]
as a temporary solution for rhel5, i think you can add `-p python26-libs` to your `script_args` and it will install both archs of python26-libs

**Comment by gtmanfred** [comment]
nevermind, looks like EXTRA_PACKAGES are installed after the main dependencies.

**Comment by gtmanfred** [comment]
```
[root@salt ~]# cat /etc/salt/cloud.deploy.d/salt-bootstrap-rhel5.sh
#!/bin/bash
set -e

yum install -y python26-libs python26.x86_64 libffi.x86_64

curl -o /tmp/boot.sh -Ls https://bootstrap.saltstack.com
bash /tmp/boot.sh "${@}"
rm -f /tmp/boot.sh
```

and
```
[root@salt ~]# cat /etc/salt/cloud.deploy.d/salt-bootstrap-rhel7.sh
#!/bin/bash
set -e

cat > /etc/yum.repos.d/salt.repo <<HERE
####################
# Enable SaltStack's package repository
[saltstack-repo]
name=SaltStack repo for RHEL/CentOS 7
baseurl=https://repo.saltstack.com/yum/rhel7
enabled=1
gpgcheck=0
gpgkey=https://repo.saltstack.com/yum/rhel7/SALTSTACK-GPG-KEY.pub
HERE

curl -o /tmp/boot.sh -Ls https://bootstrap.saltstack.com
bash /tmp/boot.sh "${@}"
rm -f /tmp/boot.sh
```

**Comment by jfindlay** [comment]
@gtmanfred, thanks for the report.

@dmurphy18

**Comment by dmurphy18** [comment]
@gtmanfred With packages for Redhat 7, some of the base dependencies are made available from the base sub-directory, for those systems which have turned off extras/updates, and sometimes even base packages from Redhat.  In  these cases the normally OS supplied dependencies are provided from the base subdirectory, and this necessitates installing the key RPM-GPG-KEY-CentOS-7 for these Centos 7 supplied dependencies.  A repo install rpm is under development to simplify this issue of not installing both keys (Salt and Centos 7).  I shall take a look into the Redhat 5 issue with libs.

**Comment by dmurphy18** [comment]
@gtmanfred Also note that on Redhat 5, a different key is used (SALTSTACK-EL5-GPG-KEY.pub).  This is due to limitations of the Redhat 5 platform and the signing methods it supports.

**Comment by gtmanfred** [comment]
Yes, the key is no the problem on RHEL 5, it is a problem with the python26-libs not being pulled in correctly based on versions.

In rhel5, if you had both versions available, it could be pulling in python26-libs as well, and on an x86_64 machine, if it pulls in the i686 package, it does not grab the python26-libs version.

**Comment by RonWilliams** [comment]
+1 
Also encountering this issue where a previously working bootstrap on RHEL7 is no longer functional.

**Comment by RonWilliams** [comment]
I found that this issue is similar to https://github.com/saltstack/salt/issues/28144.

**Comment by vutny** [comment]
I believe this is already fixed with latest repo updates and in `salt-bootstrap` develop branch. I've got everything work smoothly on both RHEL5 and RHEL7:
* https://gist.github.com/vutny/d710993c13bca9f53fba
* https://gist.github.com/vutny/54a2d09b2b8f8f8b6fd5

Some dependencies are served from `saltstack` official repository, some from `epel`.
@RonWilliams Does it work for you?

**Comment by gtmanfred** [comment]
It appears that this has been fixed.

I am closing it.

**Comment by RonWilliams** [comment]
@gtmanfred This issue was resolved as linked by @vutny. Thanks for the followup.

**Comment by sivann** [comment]
Still does not work.

**Comment by vutny** [comment]
@sivann Salt team released stable bugfix version few days ago, so there is a probability that something could be broken.

Any details about your environment and errors you've seen would be appreciated.

**Comment by sivann** [comment]
@vutny I tried both stable and dev versions. OS was RedHat 7 fully updated. I installed by adding gpgcheck=0 on salt's repo. First I tried the dev version, then the stable which also failed As I don't have access to a 2nd similar OS I cannot test from scratch the stable version on a "virgin" redhat.

**Comment by vutny** [comment]
@sivann I suggest you to open a separate issue at https://github.com/saltstack/salt-bootstrap/issues and provide as much details as you can, such as how you run bootstrap script and paste its output.

Your issue looks very similar to this, but I have just a guess that you got no repositories configured/enabled/**subscribed** except of repo.saltstack.com, and all the packages are served from there.

Try this:
```
repoquery -a --qf "%-20{repoid} %{name}" | grep libyaml
```

If I'm right, you need to add the GPG key to your `/etc/yum.repos.d/repo-saltstack-el7.repo` file:

```
[repo-saltstack-el7]
name=SaltStack EL7 Repo
baseurl=https://repo.saltstack.com/yum/rhel7/
skip_if_unavailable=True
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/rhel7/SALTSTACK-GPG-KEY.pub
       https://repo.saltstack.com/yum/rhel7/base/RPM-GPG-KEY-CentOS-7
enabled=1
enabled_metadata=1
```

and try again.
Current version of Salt bootstrap doesn't add the key for those base CentOS packages.

**Comment by gwaters** [comment]
I am having this issue as well trying to use the latest RHEL7 bootstrap. I have tried both the gpg keys that vutny suggests, but no luck. I tried downloading the gpgkeys and manually installing them, still the same error. See the gist here:
https://gist.github.com/gwaters/4c1760ec9f6eb2eed9e6

I did update my salt-stack.repo file like vutny suggested no luck.
Like Yamamushi, my build is in AWS.

**Comment by vutny** [comment]
@gwaters 
I've tried to install Salt release 2015.8.7 packages manually with all subscriptions disabled. Despite the missing `systemd-python` package (which I need to install before Salt), it went OK. But that's another story, I've already opened a separate issue.

**Comment by gwaters** [comment]
@vutny I opened this issue, which is nearly the same as this one, and Justin helped me figure out salt-cloud -u wasnt working as expected.
https://github.com/saltstack/salt/issues/31193#issuecomment-185480658
